src/lib/structures/ty.ml

```diff
@@ -311,18 +261,64 @@ let rec equal t1 t2 =
 
   | _ -> false
 
-(*** matching with a substitution mechanism ***)
-module M = Util.MI
-type subst = t M.t
+module Subst = struct
+  type subst = t TvMap.t
+
+  let id = TvMap.empty
+
+  let is_id sbt = TvMap.is_empty sbt
+
+  let eval sbt tv =
+    match TvMap.find tv sbt with
+    | ty -> ty
+    | exception Not_found -> Tvar tv
+
+  let update tv ty sbt =
+    match ty with
+    | Tvar tv' when DE.Ty.Var.equal tv tv' -> sbt
```

Comment:
If we consider a substitution as a total map from type variables into types, this is the default binding.

Reply:
Ah yes sorry, I mixed things up with rewrite rules.

Reply:
(Although given the way we use substitutions I think that they are intended to be destructive so that the binding `x → x` would always be incorrect, but that is a meta-property)